### PR TITLE
[DomCrawler] Add argument `$normalizeWhitespace` to `Crawler::innerText()` and make it return the first non-empty text

### LINKED ISSUE #48682: DomCrawler component Crawler::innerText() doesn't get text after a child node
### Symfony version(s) affected

5.4

### Description

The test case for this method basically is: 

```html
<div id="complex-element">
    Parent text
    <span>Child text</span>
</div>
```

Getting `$crawler->text()` then returns `Parent text Child text` and `$crawler->innerText()` returns `Parent text`. So far, so good.

If I don't misunderstand anything, in case of:

```html
<div id="complex-element">
    Parent text
    <span>Child text</span>
    More parent text
</div>
```

the return value of `$crawler->innerText()` should be `Parent text More parent text`. But it's still just `Parent text`.

I'll provide a pull request to fix this in a minute.

### How to reproduce

In the `Tests/AbstractCrawlerTest.php` test file, adapt the HTML in the `createTestCrawler()` method from: 

```html
<div id="complex-element">
    Parent text
    <span>Child text</span>
</div>
```

to

```html
<div id="complex-element">
    Parent text
    <span>Child text</span>
    More parent text
</div>
```

And the `testInnerText()` method accordingly to:
```php
public function testInnerText()
{
    self::assertCount(1, $crawler = $this->createTestCrawler()->filterXPath('//*[@id="complex-element"]'));

    self::assertSame('Parent text Child text More parent text', $crawler->text());
    self::assertSame('Parent text More parent text', $crawler->innerText());
}
```

### Possible Solution

https://github.com/symfony/symfony/pull/48684 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by otsch**
This is a new PR instead of https://github.com/symfony/symfony/pull/48684 with target branch 6.3 as requested.

| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #48682
| License       | MIT

### CODE REVIEW

**Review by fabpot — CHANGES_REQUESTED** [review]
You need to follow the instructions I've given in the other PR to avoid having a BC break.

**Comment by otsch** [comment]
> You need to follow the instructions I've given in the other PR to avoid having a BC break.

@fabpot Ah OK, you mean like this? https://github.com/symfony/symfony/pull/48940/commits/88a25758d5a65f8ed430c630e7ffdb0b876a5182

**Comment by otsch** [comment]
@fabpot sorry I just ammended the commit: https://github.com/symfony/symfony/pull/48940/commits/9c08134769a558074bcb3c5ac20eda57ffd663a4 to have it exactly as you explained. But the fabbot.io action doesn't like the `@param` phpdoc: https://fabbot.io/report/symfony/symfony/48940/9c08134769a558074bcb3c5ac20eda57ffd663a4 
I also tried adding the text `parameter not implemented yet` like it's also in other places with commented method arguments, but it still complains 🤔

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 1]
```diff
@@ -553,18 +553,28 @@ public function text(string $default = null, bool $normalizeWhitespace = true):
...
+            return $this->normalizeWhitespace($text);
         }
 
         return $text;
     }
 
     /**
      * Returns only the inner text that is the direct descendent of the current node, excluding any child nodes.
+     *
+     * @param bool $normalizeWhitespace parameter not implemented yet
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DomCrawler/Crawler.php]: `@param bool $normalizeWhitespace Whether whitespaces should be trimmed and normalized to single spaces`

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 2]
```diff
@@ -553,18 +553,28 @@ public function text(string $default = null, bool $normalizeWhitespace = true):
...
-        return $this->filterXPath('.//text()')->text();
+        $normalizeWhitespace = 1 <= \func_num_args() ? func_get_arg(0) : true;
+
+        foreach ($this->getNode(0)->childNodes as $childNode) {
+            if (\XML_TEXT_NODE === $childNode->nodeType && '' !== trim($childNode->nodeValue)) {
+                return $normalizeWhitespace ? $this->normalizeWhitespace($childNode->nodeValue) : $childNode->nodeValue;
+            }
+        }
+
+        return '';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DomCrawler/Crawler.php:583]: `return $normalizeWhitespace ? '' : $this->filterXPath('.//text()')->text(false);`?

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 1 above)
Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php]: 👍🏻 changed it. Still the fabbot.io github action doesn't like this phpdoc at all: https://fabbot.io/report/symfony/symfony/48940/83918eb8752ba86a76c34d4ecae52e525aa97e92

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 2 above)
Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php:583]: I'd leave the `return '';`, if that's OK. Your suggestion would return `Child text` for example, for this case: `<div class="element"><span>Child text</span>  <span>Another child</span></div>` which is text from within a child node, which contradicts the methods description imo.

I've also added two more test cases where the selected element has text only within child nodes: https://github.com/symfony/symfony/pull/48940/commits/83918eb8752ba86a76c34d4ecae52e525aa97e92#diff-efbc92395bfec1b1b53ab864d9fad40ef9bfa84e08591e56ae38b3be2e83f50aR351 

I hope it's OK how I've changed it to use a dataprovider.

Comment (nicolas-grekas) [reply src/Symfony/Component/DomCrawler/Crawler.php:583]: When $normalizeWhitespace is false, I think we should return the untrimmed inner text, that's why I made this suggestion. Or should we remove the trim when comparing in the loop in this case?

Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php:583]: OK, how about a second loop without the trim?

```php
foreach ($this->getNode(0)->childNodes as $childNode) {
    if (\XML_TEXT_NODE === $childNode->nodeType && '' !== trim($childNode->nodeValue)) {
        return $normalizeWhitespace ? $this->normalizeWhitespace($childNode->nodeValue) : $childNode->nodeValue;
    }
}

if (!$normalizeWhitespace) {
    foreach ($this->getNode(0)->childNodes as $childNode) {
        if (\XML_TEXT_NODE === $childNode->nodeType) {
            return $childNode->nodeValue;
        }
    }
}

return '';
```

Removing the trim in the first loop would get you just a space character in a case like this:
```html
<div class="element"> <span>Child</span> Parent text </div>
```

I think, even with the normalize whitespace parameter set to false, this shouldn't be the result.
Should I add a second loop?

Comment (nicolas-grekas) [reply src/Symfony/Component/DomCrawler/Crawler.php:583]: if one opts out from whitespace normalization, it means whitespaces have a meaning for them, so yes, the space should be returned.

Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php:583]: OK 👍🏻, I've added the second loop: https://github.com/symfony/symfony/pull/48940/commits/f8b1d53a79b7ba44edd719ae812c77ca998a121c

Comment (nicolas-grekas) [reply src/Symfony/Component/DomCrawler/Crawler.php:583]: thanks, I've updated the code to really return the first non-empty text, whitespaces included.

Comment (otsch) [reply src/Symfony/Component/DomCrawler/Crawler.php:583]: Ah, thanks...that's way better! 😅👍🏻

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @otsch.

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 2 above)
Comment (Bilge) [reply src/Symfony/Component/DomCrawler/Crawler.php:583]: As the author of the original implementation, I would generally agree that your interpretation of the method description is correct, however it should be noted this is nevertheless a breaking change.

**Pull request merged**